membership-attribute-service/app/services/PaymentFailureAlerter.scala

```diff
@@ -0,0 +1,85 @@
+                          paymentMethodGetter: PaymentMethodId => Future[String \/ PaymentMethodResponse])(implicit ec: ExecutionContext) : Future[Option[String]] = {
+
+    def expectedAlertText: Future[Option[String]] = {
+      val formatter = DateTimeFormat.forPattern("d MMMM yyyy").withLocale(Locale.ENGLISH)
+      val paymentMethodLatestDateFormatted: Future[String \/ String] = accountSummary.defaultPaymentMethod.map(_.id) match {
+        case Some(id) => {
+          val paymentMethod: Future[Disjunction[String, PaymentMethodResponse]] = paymentMethodGetter(id) fallbackTo Future.successful(\/.left("Failed to get payment method"))
+          paymentMethod map { paymentMethodDisjunction: Disjunction[String, PaymentMethodResponse] => paymentMethodDisjunction map { pm => pm.lastTransactionDateTime.toString(formatter)}}
+        }
+        case None => Future.successful(\/.left("No payment method id and so no payment method!"))
+      }
+
+      paymentMethodLatestDateFormatted.map { formattedDateDisjunction: Disjunction[String, String] =>
+        val alertTextWithDate = formattedDateDisjunction map { date: String =>
+          val productName = subscription.plan.name
+          s"Our attempt to take payment for your $productName membership failed on $date. Please check below that your card details are up to date."
+        }
+        alertTextWithDate.toOption
+      }
+    }
+
+    alertAvailableFor(accountObject(accountSummary), subscription, paymentMethodGetter) flatMap  { shouldShowAlert: Boolean =>
+      expectedAlertText.map { someText => shouldShowAlert.option (someText).flatten }
+    }
+
+  }
+
+  def alertAvailableFor(
+    account: AccountObject, subscription: Subscription[AnyPlan],
+    paymentMethodGetter: PaymentMethodId => Future[String \/ PaymentMethodResponse]
+  )(implicit ec: ExecutionContext): Future[Boolean] = {
+
+    def creditCard(paymentMethodResponse: PaymentMethodResponse) = paymentMethodResponse.paymentMethodType == "CreditCardReferenceTransaction" || paymentMethodResponse.paymentMethodType == "CreditCard"
+
+    val stillFreshInDays = 27
+    def recentEnough(lastTransationDateTime: DateTime) = lastTransationDateTime.plusDays(stillFreshInDays).isAfterNow
+    val isActionablePaymentGateway = account.PaymentGateway.exists(gw => gw == StripeUKMembershipGateway || gw == StripeAUMembershipGateway)
+
+    val userInPaymentFailure: Future[\/[String, Boolean]] = {
+      if(account.Balance > 0 && account.DefaultPaymentMethodId.isDefined && isActionablePaymentGateway) {
+        val eventualPaymentMethod: Future[\/[String, PaymentMethodResponse]] = paymentMethodGetter(account.DefaultPaymentMethodId.get)
+
+        eventualPaymentMethod map { maybePaymentMethod: \/[String, PaymentMethodResponse] =>
+          maybePaymentMethod.map { pm: PaymentMethodResponse =>
+            creditCard(pm) &&
+              pm.numConsecutiveFailures > 0 &&
+              recentEnough(pm.lastTransactionDateTime)
```

Comment:
so we don't actually check the dates or status of the subscription, we just care about the account balance and status, which sounds sensible to me.  Matt had a query whether we were pulling in people we shouldn't, but this looks reasonable to me.

Reply:
Yeah I think me and @lmath discussed this, and decided that if you cancelled in the future but still owed us money for a recent invoice, it was probably still correct to show the banner